### PR TITLE
fix: handle read extent with no registered data chunks

### PR DESCRIPTION
**Pull request by adammoody**
Resolves https://github.com/LLNL/UnifyFS/issues/580

### Description

This handles the case where a remote server has no extent information for a read request.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted.

### CODE REVIEW

**Comment by clmendes** [comment]
@adammoody , I'm glad to report that this fix indeed fixes the error with the cchunk5 example that I had reported in issue #580.

To be clear: with this fix, and without using the UNIFYFS_CLIENT_WRITE_SYNC=1 setting, the example code presents the same behavior as with the new-margotree branch: the code goes to the end of the execution without error, but the data stored in the file has a few errors (due to the lac of the sync).  However, now, when I use UNIFYFS_CLIENT_WRITE_SYNC=1, the file ends up with the correct data!

Thus, by using UNIFYFS_CLIENT_WRITE_SYNC=1 we no longer need the H5Fflush call in the app, as designed! I verified that the example works correctly and produces correct data in the resulting file. I also tested the original HDF5 testsuite with cchunk5, and it now works fine, producing correct data as well when we have UNIFYFS_CLIENT_WRITE_SYNC=1 in the job and the original application code does NOT contain any H5Fflush!

**Comment by adammoody** [comment]
Good to hear!  Thanks for testing this so quickly, @clmendes

**Comment by adammoody** [comment]
Before merging, I should also check the corresponding code on the send side:
https://github.com/LLNL/UnifyFS/blob/c4c2faa5d5e954d59fe1279c1f3f32c319b80c6c/server/src/unifyfs_p2p_rpc.c#L318-L335

Seems like this should be symmetric.

**Review by MichaelBrim — APPROVED** [review]

**Comment by MichaelBrim** [comment]
@adammoody In the case of no extents found, I believe `ret != UNIFY_SUCCESS`, and therefore we won't try to do the margo_bulk_create on the send side. Thus, I think your fix is complete.

**Comment by adammoody** [comment]
I got a chance to trace the send path today.  Actually, it is also allocating a bulk transfer handle ``ret == UNIFYFS_SUCCESS`` even if no chunks are returned in this case.  I'll make the corresponding change on the send side.

I think it's interesting that margo lets one create a handle successfully with ``buf_sz = 0`` from the send side ``HG_BULK_READ_ONLY`` but not on the receive side ``HG_BULK_WRITE_ONLY``.  I don't know if it's unhappy because ``buf_sz = 0`` or perhaps it has something to do with the value of the buffer address we're passing in -- maybe it can't be pinned for write or something?

I think we'd want to code around this either way, since I never really trust the outcome of a ``malloc(0)``.  However, we could raise this question with the margo team.  I'm always a fan of symmetric interfaces myself, so that they would either both fail or both succeed.

**Comment by adammoody** [comment]
@MichaelBrim , thanks for looking this over.  I force pushed after making an update to the send side.  I changed the code a bit in how we fill the output param struct and free the handle.  Let me know if that looks problematic.

**Comment by MichaelBrim** [comment]
@adammoody updates seem good to me.

File: server/src/unifyfs_p2p_rpc.c [hunk 1]
```diff
@@ -315,35 +316,37 @@ static void find_extents_rpc(hg_handle_t handle)
...
         LOGERR("margo_respond() failed");
     }
-    if (out.num_locations) {
-        margo_bulk_free(bulk_resp_handle);
-    }
 
     /* free margo resources */
+    if (bulk_resp_handle != HG_BULK_NULL) {
+        margo_bulk_free(bulk_resp_handle);
+    }
```
Comment (adammoody) [inline server/src/unifyfs_p2p_rpc.c:349]: I'm not sure if it is possible for the above ``margo_bulk_create`` to fail and set ``hret`` to something other than ``HG_SUCCESS`` but still fill in the ``bulk_resp_handle`` handle value with something other than ``HG_BULK_NULL``.  And if that is possible, do we need to call ``margo_bulk_free`` or not?  I don't know.

Comment (MichaelBrim) [reply server/src/unifyfs_p2p_rpc.c:349]: I would need to check the margo code, but I assume they do the right thing and don't assign to the output handle parameter when the operation fails. The only safe thing we can do is to only call `margo_bulk_free()` when `margo_bulk_create()` was successful. That means you should use the return value from the call to the create as the if condition.

Comment (MichaelBrim) [reply server/src/unifyfs_p2p_rpc.c:349]: I checked the margo code, it just passes straight through to mercury hg_bulk_create(), which only sets the output parameter on success.

Comment (MichaelBrim) [reply server/src/unifyfs_p2p_rpc.c:349]: https://github.com/mercury-hpc/mercury/blob/v1.0.1/src/mercury_bulk.c

Comment (adammoody) [reply server/src/unifyfs_p2p_rpc.c:349]: Cool.  Thanks, @MichaelBrim

**Comment by adammoody** [comment]
@CamStan , @kathrynmohror just merged this PR

**Comment by adammoody** [comment]
@clmendes , this fix is merged into dev now.

**Pull request merged**